### ISSUE TITLE
compositionality of update functions

### ISSUE THREAD

**Issue by benanne** [benanne/Lasagne#190]
Currently, the interface of the functions in `lasagne.updates` is as follows:

```
def some_update_function(loss, list_of_params, **hyperparameters):
    ...
```

Each update function has the responsibility to call `T.grad()` on `loss`.
This is a bit redundant, but also, it makes it impossible to combine different operations. For example, some people like to combine RMSprop with Nesterov momentum. Currently the only way to do this in Lasagne would be to implement a new update function and copy-paste the relevant code. There is no clean way to combine the existing code for RMSprop and Nesterov momentum programmatically.

It may pay off to come up with a better interface to make these things more compositional, provided that it does not increase complexity too much. Basically, we'd need a set of functions that transform updates (e.g. the `momentum` function would take a set of updates and apply the momentum update rule to it). So its interface would then be:

```
def some_update_transformation_function(updates, **hyperparameters):
    ...
```

We can still provide 'shortcuts' for common combinations that provide the interface as it is currently.

I haven't really thought this through, but I think it's a valid concern (one of my colleagues alerted me to it). What are everyone's ideas on this?

**Comment by craffel** [comment]
Yes, this is something I've thought of, because many of the recently proposed stochastic optimization techniques involve the same ingredients (momentum as you said, but also exponentially-decaying first and second order moment estimates, etc).  I couldn't think of a good way to implement this so I never suggested it.

**Comment by skaae** [comment]
I had to do this for some LSTM code and for the DRAW code. 

I ended up using:

```Python
all_grads = theano.grad(cost,all_params
all_grads = step_clipping(all_grads, threshold=10, to_zero=False)
updates = adam(all_grads, all_params, learning_rate=0.0003)
```

But this is not very thought through. I know that blocks has a [compositeRule](https://github.com/bartvm/blocks/blob/9396fa5c668076693e174105e3aaf244a7a3efe9/blocks/algorithms/__init__.py#L302) which takes a list of step rules. Heres an example of adam+stepclipping in blocks:

```Python
algorithm = GradientDescent(
cost=cost,
params=params,
step_rule=CompositeRule([
StepClipping(10.),
Adam(learning_rate),
])
```

I think each steprule returns a list of updates that are then applied in order. For the above example a bunch of updates would first clip the gradients and then the adam updates would follow?

**Comment by f0k** [comment]
I ended up using:
```python
all_grads = theano.grad(cost,all_params
all_grads = step_clipping(all_grads, threshold=10, to_zero=False)
updates = adam(all_grads, all_params, learning_rate=0.0003)
```

Doesn't look too bad (except for a missing closing bracket.
;)
But this addresses another problem: generating update expressions for user-supplied gradients. Also a valid use case, but not the one that would allow compositing RMSProp and Nesterov momentum.

To solve all three, we could have the current update functions accept either a cost expression and a list of params, or list of gradients and a list of params, or an update dictionary and a list of params. The update functions should return an OrderedDict instead of a list of tuples then. (That would make a lot of things easier anyway. What do you think about that?)

The question is whether the three-types-of-input case can be handled easily. The first two wouldn't be a problem, the `all_grads = theano.grad(loss, all_params)` call would just need to be skipped:
```python
def momentum(something, all_params, learning_rate, momentum=0.9):
    if isinstance(something, list):
        all_grads = something
    else:
        all_grads = theano.grad(something, all_params)
    updates = OrderedDict()

    for param_i, grad_i in zip(all_params, all_grads):
        mparam_i = theano.shared(np.zeros(param_i.get_value().shape,
                                          dtype=theano.config.floatX),
                                 broadcastable=param_i.broadcastable)
        v = momentum * mparam_i - learning_rate * grad_i
        updates[mparam_i] = v
        updates[param_i] = param_i + v

    return updates
```

About the third one we'd need to think a bit. What would `momentum` look like if supplied with an updates dictionary that came out of RMSProp? Maybe something like this?

```python
def momentum(something, all_params, learning_rate, momentum=0.9):
    if not isinstance(something, dict):
        updates = sgd(something, all_params, learning_rate)
    else:
        updates = OrderedDict(updates)

    for param_i in all_params:
        mparam_i = theano.shared(np.zeros(param_i.get_value().shape,
                                          dtype=theano.config.floatX),
                                 broadcastable=param_i.broadcastable)
        v = momentum * mparam_i + updates[param_i] - param_i
        updates[mparam_i] = v
        updates[param_i] = param_i + v

    return updates
```

I think this would work, wouldn't it? It only makes sense for `momentum` and `nesterov_momentum`, probably, but those could be propped onto anything then (anything that produces updates of the form `param_i = param_i - foo`). The only remaining ugliness is that my proposal still requires a `learning_rate` argument although this is only used when falling back to using `sgd` updates. This could be solved by turning `momentum` and `nesterov_momentum` into functions that transform an updates dictionary rather than generating one, but that would require every user to compose this with `sgd`. Hmm. Opinions?

**Comment by f0k** [comment]
Alternative: Two extra functions.

```python
def apply_momentum(updates, params, momentum=0.9):
    updates = OrderedDict(updates)
    for param_i in all_params:
        mparam_i = theano.shared(np.zeros(param_i.get_value().shape,
                                          dtype=theano.config.floatX),
                                 broadcastable=param_i.broadcastable)
        v = momentum * mparam_i + updates[param_i] - param_i
        updates[mparam_i] = v
        updates[param_i] = param_i + v

def momentum(cost_or_grad, all_params, learning_rate, momentum=0.9):
    if not isinstance(cost_or_grad, list):
        cost_or_grad = theano.grad(cost_or_grad, all_params)
    updates = sgd(cost_or_grad, learning_rate)
    return apply_momentum(updates, momentum)
```

Backwards compatible and easy. `nesterov_momentum` would be done analogously. I'd still prefer to change all existing update functions to accept gradient lists as an alternative to cost expressions, and to return OrderedDicts.

**Comment by skaae** [comment]
I think it is cleaner, and easier to understand, if we let the user compose the udate rule. The good thing with lasagne is that most things are simple to do, but you sort of have to do it, it doesn't happen automatically. 

I would let all update rules take the gradients as input instead of the loss function. Then we can create some conditioning functions like step_clipping etc that the user can apply either on the updates or on the gradients from T.grad?

**Comment by benanne** [comment]
That's true, but we also want to make the common use cases easy. So in the interest of pragmatism I think it is worth supporting the current interface as well, even if it's not as "clean".

If we 'clean up' the interface like this step by step, and make people write a little more code every time we do it, eventually they're going to have to write twice as much, so I think we should avoid that slippery slope when we can!

**Comment by skaae** [comment]
Yes, I like the first momentum that @f0k suggested, then we'll just have to add the clipping functions etc?

**Comment by benanne** [comment]
Sounds good to me!

If anyone wants to turn this into a PR, go for it! If not, I probably will sometime next week.

**Comment by f0k** [comment]
I'll go for it, using my last proposal. I'll include changing return values from tuple lists to OrderedDict.

**Comment by benanne** [comment]
This has been addressed in #194.
